### DISCUSSION TITLE
How to run jzy3d JOGL 2.4 SWT examples

### DISCUSSION THREAD

**Discussion by Bio7** [jzy3d/jzy3d-api#252] [General]
Hello Martin,

I just evaluate some OpenGL 3D charting libraries.and was trying to embed jzy3d in an SWT app example with JOGL 2.4 RC.
I followed your maven recipe and tried several other things to make it run but could'nt get it working.

Do you have an updated maven documentation (or a link to the necessary  *.jars) for JOGL 2.4?

**Comment by jzy3d** [comment]
Hi @Bio7,

The SWT demo is inside the SWT module of Jzy3D [here](https://github.com/jzy3d/jzy3d-api/blob/master/jzy3d-native-jogl-swt/src/main/java/org/jzy3d/demos/SurfaceDemoSWT.java). You could copy paste this example and depend on

```xml
<groupId>org.jzy3d</groupId>
<artifactId>jzy3d-native-jogl-swt</artifactId>
<version>2.0.1-SNAPSHOT</version>
```

this will bring JOGL 2.4 rc4.

Note that part of the SWT jar selection is made [here](https://github.com/jzy3d/jzy3d-api/blob/master/jzy3d-native-jogl-swt/pom.xml#L53). 

The readme of this SWT module is not relevant. I should update it.

**Comment by Bio7** [comment]
Hello Martin,

thanks for the fast reply. However, do you have a complete maven file available for the SWT example?

Once again I had no success to built and execute the example.

**Reply by jzy3d** [reply]
I am writing a [standalone SWT example](https://github.com/jzy3d/jzy3d-swt-example). I am facing the below Maven issue. Is that what you encountered?

`Could not resolve dependencies for project org.jzy3d:jzy3d-swt-example:jar:2.0.1-SNAPSHOT: Failed to collect dependencies at org.jzy3d:jzy3d-native-jogl-swt:jar:2.0.1-SNAPSHOT -> org.eclipse.platform:org.eclipse.jface:jar:3.12.1 -> org.eclipse.platform:org.eclipse.swt:jar:3.105.2 -> org.eclipse.platform:org.eclipse.swt.gtk.linux.aarch64:jar:[3.105.2,3.105.2]: No versions available for org.eclipse.platform:org.eclipse.swt.gtk.linux.aarch64:jar:[3.105.2,3.105.2] within specified range -> [Help 1]
`

The strange thing is that this Linux Aarch64 should not be queried since I am on macOS + Intel CPU.

**Comment by Bio7** [comment]
Hello Martin,

if import the Maven project into Eclipse I got the following error if I try to built the project POM. I'm on Windows. So I also get the Linux error.

[INFO] Building Jzy3d SWT Example 2.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.116.100 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.105.3 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.106.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.106.1 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.106.2 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.106.3 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.107.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.108.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.109.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.110.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.111.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.112.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.113.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.114.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.114.100 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.115.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.115.100 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.116.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.117.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.118.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.swt:jar:3.119.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[WARNING] The POM for org.eclipse.platform:org.eclipse.jface:jar:3.116.100 is missing, no dependency information available
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.428 s
[INFO] Finished at: 2022-05-02T16:41:58+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project jzy3d-swt-example: Could not resolve dependencies for project org.jzy3d:jzy3d-swt-example:jar:2.0.1-SNAPSHOT: Failed to collect dependencies at org.jzy3d:jzy3d-native-jogl-swt:jar:2.0.1-SNAPSHOT -> org.eclipse.platform:org.eclipse.jface:jar:3.12.1 -> org.eclipse.platform:org.eclipse.swt:jar:3.105.2 -> org.eclipse.platform:org.eclipse.swt.gtk.linux.aarch64:jar:[3.105.2,3.105.2]: No versions available for org.eclipse.platform:org.eclipse.swt.gtk.linux.aarch64:jar:[3.105.2,3.105.2] within specified range -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException

**Reply by jzy3d** [reply]
The same mistake I encountered in my standalone demo. It is weird we all get this Linux Aarch64 jar search for either MacOS or Windows. 

@HannesWell, as you improved the SWT dependencies in Jzy3D and are more used than me integrating Jzy3D in Eclipse apps, do you have an explanation about this issue? Since your improvements I only added [OS/CPU detection in SWT build profile](https://github.com/jzy3d/jzy3d-api/blob/master/jzy3d-native-jogl-swt/pom.xml#L49) to allow [running automated tests on all OS](https://github.com/jzy3d/jzy3d-api/blob/master/.github/workflows/maven.yml#L86).

**Reply by HannesWell** [reply]
Sorry for the late reply.
I just checked it and the build of jzy3d itself (at the current master) on my Windows computer succeeds without any errors (even all default tests pass).

It is not completely clear if you have the issue when build jzy3d itself or a project that depends on it?

But the profiles look fine.

**Reply by jzy3d** [reply]
Thanks for checking.

We are having problems with an SWT app that depend on Jzy3D. Here is an example to reproduce it : https://github.com/jzy3d/jzy3d-swt-example

The problem seams to be with the SWT lib itself and I wonder if you already faced it ?

The stack above is clearer but what we get at the end is :
`Could not resolve dependencies for project org.jzy3d:jzy3d-swt-example:jar:2.0.1-SNAPSHOT: Failed to collect dependencies at org.jzy3d:jzy3d-native-jogl-swt:jar:2.0.1-SNAPSHOT -> org.eclipse.platform:org.eclipse.jface:jar:3.12.1 -> org.eclipse.platform:org.eclipse.swt:jar:3.105.2 -> org.eclipse.platform:org.eclipse.swt.gtk.linux.aarch64:jar:[3.105.2,3.105.2]: No versions available for org.eclipse.platform:org.eclipse.swt.gtk.linux.aarch64:jar:[3.105.2,3.105.2] within specified range
`

**Reply by HannesWell** [reply]
Thanks for the example, I can re reproduce the failure but I could not yet fix it, but I will continue my search.
At the moment I suspect that JFace/SWT is configured badly or at least requires some special configuration. IIRC it is tricky to set up correctly.

What I already noticed is that the jface dependency has the wrong version. It should be:
```
<dependency>
	<groupId>org.eclipse.platform</groupId>
	<artifactId>org.eclipse.jface</artifactId>
	<version>3.25.0</version>
</dependency>
```
The jface version in the jzy3d-api could also be updated. Have you considered to enable depandabot for this repository?

However all of this did not fix the issue, but I will continue my investigation (but it could take a bit).

**Comment by Bio7** [comment]
Hello Martin, any news how to successfully built the jzy3d SWT example with JOGL 2.4RC4?

**Reply by Bio7** [reply]
Hello Martin any chance that this issue will be fixed soon? 

 If not by Maven what are the minimum *.jar dependencies to run the example?
